File: CS50x/ProblemSets/C/credit.c
// Check if the card number is valid
#include <cs50.h>
#include <stdio.h>

// Length and the first two digits of each type of card
#define AMEX 15
#define MASTERCARD 16
#define VISA 13
#define AMEX_START1 34
#define AMEX_START2 37
#define MASTERCARD_LOW 51
#define MASTERCARD_HIGH 55
#define VISA_START 4

string check(long number);

// main function
int main(void)
{
    long card_number = get_long("Number: ");
    string result = check(card_number);
    printf("%s\n", result);
}

// Using Luhn’s Algorithm, also checking for length and the first two digits
string check(long number)
{
    int first_two_digits;
    int length = 0;
    int checksum = 0;

    // Get the checksum, the length and the first_two_digits
    while (number)
    {
        int last_digit = number % 10;
        if (length % 2 == 0)
            checksum += last_digit;
        else
        {
            int cur = last_digit * 2;
            if (cur >= 10)
                checksum += cur % 10 + cur / 10;
            else
                checksum += cur;
        }
        number /= 10;
        length++;
        if (number >= 10 && number <= 99)
            first_two_digits = number;
    }

    // Check
    if (checksum % 10 != 0)
        return "INVALID";
    else if (length == AMEX && (first_two_digits == AMEX_START1 || first_two_digits == AMEX_START2))
        return "AMEX";
    else if (length == MASTERCARD && first_two_digits >= MASTERCARD_LOW &&
             first_two_digits <= MASTERCARD_HIGH)
        return "MASTERCARD";
    else if ((length == VISA || length == MASTERCARD) && first_two_digits / 10 == VISA_START)
        return "VISA";
    else
        return "INVALID";
}

File: CS50x/ProblemSets/Python/credit.py
# Check if the credit card is valid
from cs50 import get_int
import re


# Constants
AMEX_LEN, MASTERCARD_LEN, VISA_LEN = 15, 16, 13
AMEX_START = "34|37"
MASTERCARD_START = "5[1-5]"
VISA_START = "4"


# Main function
def main():
    card = get_int("Number: ")
    check = check_sum(card)
    if check:
        check_pattern(str(card))
    else:
        print('INVALID')


# Luhn
def check_sum(card):
    sum = 0
    while card:
        sum += card % 10
        card //= 10
        two = (card % 10) * 2
        if two >= 10:
            sum += two % 10 + two // 10
        else:
            sum += two
        card //= 10

    if sum % 10 == 0:
        return True
    return False


# Check card pattern
def check_pattern(card):
    length = len(card)
    if length == AMEX_LEN and re.match(AMEX_START, card):
        print("AMEX")
    elif length == MASTERCARD_LEN and re.match(MASTERCARD_START, card):
        print("MASTERCARD")
    elif (length == MASTERCARD_LEN or length == VISA_LEN) and re.match(VISA_START, card):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
